### PR TITLE
[fsdp2][mp] Fix FSDP2 mixed precision reduce dtype after unfreezing params

### PR DESCRIPTION
**Pull request by anshul-si**
**Summary:** fixes https://github.com/pytorch/pytorch/issues/186998.  FSDP2 mixed precision initializes a param group’s `_orig_dtype` and `_reduce_dtype` during the first forward. If that first forward happens
  while all params are frozen, there are no trainable params to derive dtype metadata from, so both fields were left as `None`. After later unfreezing, gradient reduction could then fall back to the bf16 compute dtype instead of the requested fp32 reduce dtype.

This updates dtype resolution so that when no params are currently trainable, FSDP2 falls back to the group’s floating-point params if their original and reduce dtypes are uniform. Mixed all-frozen groups still preserve the previous behavior by resolving to `None`.

**Test Case**
1. pytest test/distributed/_composable/fsdp/test_fully_shard_mixed_precision.py -k test_reduce_dtype_after_frozen_first_forward

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #187376

### CODE REVIEW

File: test/distributed/_composable/fsdp/test_fully_shard_mixed_precision.py [hunk 1]
```diff
@@ -254,6 +254,54 @@ def assert_fn(output: torch.Tensor):
             self.assertEqual(fsdp_loss, ref_loss)
             check_sharded_parity(self, ref_model, model)
 
+    @skipIfRocmVersionLessThan((7, 0))
+    @skip_if_lt_x_gpu(2)
+    @requires_nccl_version((2, 10), "Need NCCL 2.10+ for bf16 collectives")
```
Comment (weifengpy) [inline test/distributed/_composable/fsdp/test_fully_shard_mixed_precision.py]: the test is using `reduce_dtype=fp32`, not bf16 - is `@requires_nccl_version` really necessary ?

File: torch/distributed/fsdp/_fully_shard/_fsdp_param_group.py [hunk 2]
```diff
@@ -277,22 +277,33 @@ def _init_mp_dtypes(self) -> None:
         trainable_params: list[FSDPParam] = [
             p for p in self.fsdp_params if p.sharded_param.requires_grad
         ]
-        orig_dtypes = {p.orig_dtype for p in trainable_params}
-        reduce_dtypes = {p.reduce_dtype for p in trainable_params}
-        if len(trainable_params) > 0 and len(orig_dtypes) != 1:
+        if trainable_params:
+            params_for_dtype = trainable_params
+            should_assert_uniform_dtypes = True
```
Comment (weifengpy) [inline torch/distributed/fsdp/_fully_shard/_fsdp_param_group.py]: `trainable_params` might be able to express the same thing to `should_assert_uniform_dtypes`. It's a style-ish nit to keeping diff minimal

```
diff --git a/torch/distributed/fsdp/_fully_shard/_fsdp_param_group.py b/torch/distributed/fsdp/_fully_shard/_fsdp_param_group.py
  --- a/torch/distributed/fsdp/_fully_shard/_fsdp_param_group.py
  +++ b/torch/distributed/fsdp/_fully_shard/_fsdp_param_group.py
  @@ -277,25 +277,29 @@ class FSDPParamGroup:
           trainable_params: list[FSDPParam] = [
               p for p in self.fsdp_params if p.sharded_param.requires_grad
           ]
  -        orig_dtypes = {p.orig_dtype for p in trainable_params}
  -        reduce_dtypes = {p.reduce_dtype for p in trainable_params}
  -        if len(trainable_params) > 0 and len(orig_dtypes) != 1:
  -            # Models may have no grad params
  +        params_for_dtype = trainable_params or [
  +            p for p in self.fsdp_params if p.orig_dtype.is_floating_point
  +        ]
  +        orig_dtypes = {p.orig_dtype for p in params_for_dtype}
  +        reduce_dtypes = {p.reduce_dtype for p in params_for_dtype}
  +        if trainable_params and len(orig_dtypes) != 1:
               raise AssertionError(
                   f"FSDP expects uniform original parameter dtype but got {orig_dtypes}"
               )
  -        self._orig_dtype = next(iter(orig_dtypes)) if trainable_params else None
  -        if len(trainable_params) > 0 and len(reduce_dtypes) != 1:
  +        if trainable_params and len(reduce_dtypes) != 1:
               # This can be relaxed if we issue one reduce-scatter per reduce
               # dtype (but we would need a way for users to specify multiple
               # reduce dtypes)
               raise AssertionError(
                   f"FSDP expects uniform reduce dtype but got {reduce_dtypes}"
               )
  -        self._reduce_dtype = next(iter(reduce_dtypes)) if trainable_params else None
  +        dtype_sets_are_uniform = len(orig_dtypes) == 1 and len(reduce_dtypes) == 1
  +        self._orig_dtype = next(iter(orig_dtypes)) if dtype_sets_are_uniform else None
  +        self._reduce_dtype = next(iter(reduce_dtypes)) if dtype_sets_are_uniform else None
```

Comment (weifengpy) [reply torch/distributed/fsdp/_fully_shard/_fsdp_param_group.py]: for functional gap, codex shows me a corner case - i don't mind a following up PR for corner case though

```
if the model is fully frozen on first forward and has mixed floating parameter dtypes, the patch still leaves
  _orig_dtype / _reduce_dtype as None. If a uniform subset is later unfrozen, reduction can still happen in grad dtype rather than the requested
  explicit reduce_dtype.
```

**Review by weifengpy — APPROVED** [review]

File: test/distributed/_composable/fsdp/test_fully_shard_mixed_precision.py (hunk 1 above)
Comment (anshul-si) [reply test/distributed/_composable/fsdp/test_fully_shard_mixed_precision.py]: I'll get rid of it because since there's mixed dtypes, all_gather uses uint8 instead of bf16

File: torch/distributed/fsdp/_fully_shard/_fsdp_param_group.py (hunk 2 above)
Comment (anshul-si) [reply torch/distributed/fsdp/_fully_shard/_fsdp_param_group.py]: I will address this in a separate pr, i think if we want this functionality, we need to determine it before each reduce_scatter instead of just relying on defaults
